File: mirror/tier2-dart/showcase/perf_memoization.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Optimization
// Module: perf_memoization.dart

Function memoize(Function func) {
  final cache = <String, dynamic>{};
  return (List<dynamic> args) {
    final key = args.toString();
    if (!cache.containsKey(key)) {
      cache[key] = Function.apply(func, args);
    }
    return cache[key];
  };
}

int expensiveComputation(int n) {
  int result = 0;
  for (int i = 0; i < n; i++) {
    result += i * i;
  }
  return result;
}

void main() {
  final customMemoized = memoize((int x) {
    int total = 0;
    for (int i = 0; i < x; i++) {
      total += i;
    }
    return total;
  });
  
  final result = <String, dynamic>{
    'expensive5': expensiveComputation(5),
    'expensive10': expensiveComputation(10),
    'customMemo100': customMemoized([100]),
    'customMemo200': customMemoized([200]),
    'memoizationActive': true,
  };
  print(result);
}

File: mirror/tier2-ruby/showcase/perf_memoization.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Optimization
# Module: perf_memoization.rb
# Purpose: Demonstrate generic memoization

def memoize(&block)
  cache = {}
  -> (*args) do
    key = args
    cache[key] ||= block.call(*args)
  end
end

def expensive_computation(n)
  result = 0
  n.times { |i| result += i * i }
  result
end

custom_memoized = memoize do |x|
  total = 0
  x.times { |i| total += i }
  total
end

result = {
  expensive_5: expensive_computation(5),
  expensive_10: expensive_computation(10),
  custom_memo_100: custom_memoized.call(100),
  custom_memo_200: custom_memoized.call(200),
  memoization_active: true,
}

puts result.inspect
